middleware/generate.go

```diff
@@ -0,0 +1,520 @@
+	}
+
+	return found.With, true
+}
+
+// Add injects the middleware to the relative position of the middleware group.
+//
+// Add never returns an error. It used to for duplicate phases but this
+// behavior has since been removed as part of a performance optimization. The
+// return value from Add can be ignored.
+func (s *{{.Phase}}Step) Add(m {{.Phase}}Middleware, pos RelativePosition) error {
+	if s.head == nil {
+		s.head = &decorated{{.Phase}}Handler{nil, m}
+		s.tail = s.head
+		return nil
+	}
+
+	if pos == Before {
+		s.head = &decorated{{.Phase}}Handler{s.head, m}
+	} else {
+		tail := &decorated{{.Phase}}Handler{nil, m}
+		s.tail.Next = tail
+		s.tail = tail
+	}
+
+	return nil
+}
+
+// Insert injects the middleware relative to an existing middleware ID.
+// Returns error if the original middleware does not exist, or the middleware
+// being added already exists.
+func (s *{{.Phase}}Step) Insert(m {{.Phase}}Middleware, relativeTo string, pos RelativePosition) error {
+	found, prev := s.get(relativeTo)
+	if found == nil {
+		return fmt.Errorf("not found: %s", m.ID())
+	}
+
+	if pos == Before {
+		if prev == nil { // at the front
+			s.head = &decorated{{.Phase}}Handler{s.head, m}
+		} else { // somewhere in the middle
+			prev.Next = &decorated{{.Phase}}Handler{found, m}
+		}
+	} else {
+		if found.Next == nil { // at the end
+			tail := &decorated{{.Phase}}Handler{nil, m}
+			s.tail.Next = tail
+			s.tail = tail
+		} else { // somewhere in the middle
+			found.Next = &decorated{{.Phase}}Handler{found.Next, m}
+		}
+	}
+
+	return nil
+}
+
+// Swap removes the middleware by id, replacing it with the new middleware.
+// Returns the middleware removed, or error if the middleware to be removed
+// doesn't exist.
+func (s *{{.Phase}}Step) Swap(id string, m {{.Phase}}Middleware) ({{.Phase}}Middleware, error) {
+	found, _ := s.get(id)
+	if found == nil {
+		return nil, fmt.Errorf("not found: %s", m.ID())
+	}
+
+	swapped := found.With
+	found.With = m
+	return swapped, nil
+}
+
+// Remove removes the middleware by id. Returns error if the middleware
+// doesn't exist.
+func (s *{{.Phase}}Step) Remove(id string) ({{.Phase}}Middleware, error) {
+	found, prev := s.get(id)
+	if found == nil {
+		return nil, fmt.Errorf("not found: %s", id)
+	}
+
+	if s.head == s.tail { // it's the only one
+		s.head = nil
+		s.tail = nil
+	} else if found == s.head { // at the front
+		s.head = s.head.Next.(*decorated{{.Phase}}Handler)
+	} else if found == s.tail { // at the end
+		prev.Next = nil
+		s.tail = prev
+	} else {
+		prev.Next = found.Next // somewhere in the middle
+	}
+
+	return found.With, nil
+}
+
+// List returns a list of the middleware in the step.
+func (s *{{.Phase}}Step) List() []string {
+	var ids []string
+	for h := s.head; h != nil; {
+		ids = append(ids, h.With.ID())
+		if h.Next == nil {
+			break
+		}
+
+		// once executed, tail.Next of the list will be set to an
+		// *{{.PhaseLowercase}}WrapHandler, make sure to check for that
```

Comment:
why do we need to check for this explicitly?

Reply:
See Tianyi's comments from before, the loop would have either hung or panicked on the type assert

Reply:
got it